Add support for resumption tokens

chrispenny opened the pull request:
## Resumption Token

Resumption Tokens are a form of pagination, however, they also contain a level of validation.

Each Resumption Token should represent a specific request, including whatever filters might have been applied as part of that request, as well as representing a particular "page" in the Paginated List.

The goal is to increase reliability of pagination by making sure that each requested "page" came from a request containing the expected filters. EG: You can't send an unfiltered request for OAI Records, see that there are 10 pages, and then decide to request page=10 with some filters now applied. The Token itself would be aware that a different filter has been applied, and it would be invalid.

## Consistency in config names

I realised that we had some snake case and camel case names. Switched everything to snake case to be consistent with SS framework.

https://github.com/silverstripe-terraformers/open-archive-initiative/pull/11/commits/09b0b388eeab4357fe42b62287e28cb587c50933

chrispenny committed:
Add support for resumption tokens

chrispenny committed:
Update configuration names for consistency

MelissaWu-SS reviewed (changes requested):
Awesome! I tested it on my local for the new support for `resumptionToken`, got 5 results per request, and there is a `resumptionToken` element in the result if there are more records.

I just noticed that there is a rule for the `resumptionToken` element when I read the [doc](openarchives.org/OAI/openarchivesprotocol.html#FlowControl).
![Screen Shot 2022-04-04 at 11 48 45 AM](https://user-images.githubusercontent.com/81267819/161454292-aefbde42-7287-4249-9125-9f7c948f6b2c.png)
I am wondering if we need to implement this? As I found there is no empty `resumptionToken` element in the last response. 👀

Other than that, all looks great to me! 🌟

chrispenny commented:
Nice spotting, @MelissaWu-SS! Thanks for picking that up.

chrispenny committed:
Add support for expiring Resumption Tokens

chrispenny commented:
@MelissaWu-SS I've added the empty Resumption Token for the "last page" of a paginated response.

I was also originally **not** going to bother with expiring Resumption Tokens, however, I've had a change of heart. I've decided that I don't like the idea of Resumption Tokens "living forever".

I've added some new functionality, which is the ability for us to expire our Tokens. The default is 1 hour.
https://github.com/silverstripe-terraformers/open-archive-initiative/pull/11/commits/039fa8f099b7b93f5fc5ef7c46c7bd05aa32669a

Easiest way to test the token expiry is to get your first response and grab the Resumption Token from it.
![Screen Shot 2022-04-04 at 1 31 19 PM](https://user-images.githubusercontent.com/505788/161459813-a6040ab3-4757-4198-82a6-04c13cc0357b.png)

Take it to a Base 64 decoder/encoder (https://www.base64encode.org/), and decode it.
![Screen Shot 2022-04-04 at 1 28 04 PM](https://user-images.githubusercontent.com/505788/161459960-1787808f-a432-4d06-9c2f-cf64d7599fd4.png)

Change the expiry date to "an hour earlier", re-encode it, and use that as the Resumption Token.
![Screen Shot 2022-04-04 at 1 28 30 PM](https://user-images.githubusercontent.com/505788/161459952-e5fd8613-8f27-44e8-9e51-250b50a9ba77.png)

MelissaWu-SS commented on src/Documents/ListRecordsDocument.php:
```diff
@@ -32,4 +33,20 @@ public function processOaiRecords(DataList $oaiRecords): void
+        $resumptionTokenElement = $this->findOrCreateElement('resumptionToken', $listRecordsElement);
+
+        $resumptionTokenElement->nodeValue = $resumptionToken;
+
```
```suggestion
        if (!$resumptionToken) {
            return;
        }
```
Can we please add this for the empty resumption token scenario? 👀
I just realised it would cause an error in the decode process(`$decode = base64_decode($resumptionToken, true);`) if we set an empty resumption token in the last page of the response. Looks `base64_decode` are not able to decode an empty string.


![Screen Shot 2022-04-04 at 2 30 40 PM](https://user-images.githubusercontent.com/81267819/161464304-f8214666-7c22-4834-90e9-aea8b740d570.png)

chrispenny replied:
Thanks, @MelissaWu-SS!

Fixed at the source of the issue:
https://github.com/silverstripe-terraformers/open-archive-initiative/pull/11/commits/fad9151d25d08bdccce8d341e3c8172e231956ca

MelissaWu-SS reviewed (changes requested):
Sweet! Thanks for the suggestion around manually making the token expired! Otherwise I will probably wait for one hour😀

Just one minor change, other than that, the adding expiry attribute and empty resumption token element are looking great to me!🌟

chrispenny committed:
Fix for empty resumption token
